src/test/java/org/assertj/core/test/TypeCanonizer.java

```diff
@@ -0,0 +1,127 @@
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ * <p>
+ * Copyright 2012-2017 the original author or authors.
+ */
+package org.assertj.core.test;
+
+import com.google.common.reflect.TypeResolver;
+
+import java.io.InputStream;
+import java.lang.reflect.GenericArrayType;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.lang.reflect.WildcardType;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * A Type canonizar that helps with the resolving of a {@link Type} so it can be compared to a similar one
+ * considering generics.
+ * @author Filip Hrisafov
+ * @author Clement Mathieu
+ */
+public class TypeCanonizer {
+
+  /**
+   * Returns a canonical form of {@code initialType} by replacing all {@link TypeVariable} by {@link Class}
+   * instances.
+   * <p>
+   * <p>
+   * Such a canonical form allows to compare {@link ParameterizedType}s, {@link WildcardType}(s),
+   * {@link GenericArrayType}(s), {@link TypeVariable}(s).
+   * </p>
+   */
+  public static Type canonize(Type initialType) {
+    if (doesNotNeedCanonization(initialType)) {
+      return initialType;
+    }
+
+    ReplacementClassSupplier replacementClassSupplier = new ReplacementClassSupplier();
+    TypeResolver typeResolver = new TypeResolver();
+
+    for (TypeVariable typeVariable : findAllTypeVariables(initialType)) {
+      typeResolver = typeResolver.where(typeVariable, replacementClassSupplier.get());
+    }
+
+    return typeResolver.resolveType(initialType);
+  }
+
+  private static boolean doesNotNeedCanonization(Type type) {
+    return !(type instanceof ParameterizedType || type instanceof GenericArrayType || type instanceof WildcardType
+             || type instanceof TypeVariable);
+  }
+
+  /**
+   * @param type the type for which all type variables need to be extracted
+   * @return all {@code type}'s {@link TypeVariable}
+   */
+  private static Set<TypeVariable> findAllTypeVariables(Type type) {
+    Set<TypeVariable> typeVariables = new LinkedHashSet<>();
+    findAllTypeVariables(typeVariables, type);
+    return typeVariables;
+  }
+
+  /**
+   * Adds all {@code type}'s {@link TypeVariable} to {@code typeVariables}
+   *
+   * @param typeVariables that need to be populated
+   * @param types the types for which the {@link TypeVariable}(s) need to be extracted
+   */
+  private static void findAllTypeVariables(Set<TypeVariable> typeVariables, Type... types) {
```

Comment:
rename to `populateAllTypeVariables`, when I read `find` my brain says "oh, no side effects" ;-)